### PR TITLE
Pbi 449 - markdown support and expansion panels in Help dialog

### PR DESCRIPTION
**Pull request by hrishikeshpaul**
Target issue #449 

Let me know if this is the workflow you were looking for - I'm not sure if I did it the right way.

**Workflow**
- Read the markdown as a string.
- Process it by splitting it by `#` (h1) tags.
- Create an object that has the title and content.
- Use this to display it on the expansion-panel component.

**Notes**:
- I'm not sure where the `documentation.md` file goes.
- Any specs for the styling of the expansion panels?
- There is this linting error I'm not being able to get rid off ```ccf-ui/projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts:23:40
ERROR: 23:40  no-unsafe-any  Unsafe use of expression of type 'any'.``` Any ideas?

### CODE REVIEW

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { DocumentationContent } from '../../../core/models/documentation';
+import * as documentation from 'raw-loader!./documentation.md';
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts:4]: That works, but I'd rather use HTTPService to load from assets on the fly

File: projects/ccf-rui/src/app/modules/info/info-panel/info-panel.module.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { NgModule, SecurityContext } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { InfoPanelComponent } from './info-panel.component';
+import { MarkdownModule } from 'ngx-markdown';
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-panel/info-panel.module.ts:4]: You need to add ngx-markdown to package.json/package-lock.json (build is failing)

**Review by bherr2 — CHANGES_REQUESTED** [review]
Looks pretty good. Would like to see more tests and a couple fixes.

File: projects/ccf-rui/src/app/modules/info/info-panel/info-panel.module.ts (hunk 2 above)
Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-panel/info-panel.module.ts:4]: Ah I wonder why this didn't happen. I did but the `--save` tag. I'll fix it.

**Comment by bherr2** [comment]
Sorry, just noticed the notes in the original comment.

* `documentation.md` should go in the rui's assets folder (in a subfolder called docs and actually let's call documentation.md `README.md` so that on github it displays nicely there. I say put a subfolder just in case we add images or other assets to a future expanded documentation).
* Just use the standard styling for expansion panels (use your best judgement. Leonard will review the whole site at the end of the sprint to iron out any styling issues).
* The last thing should go away when switching to HTTPService to fetch the md at runtime.

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import { Injectable } from '@angular/core';
 import { BehaviorSubject } from 'rxjs';
 import { DocumentationContent } from '../../../core/models/documentation';
-import * as documentation from 'raw-loader!./documentation.md';
+import { HttpClient } from '@angular/common/http';
+
+const documentationURL = './../../../../assets/documentation.md';
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts:6]: URL should be `assets/docs/readme.md`

Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts]: Yep, I just saw this comment after I pushed

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { DocumentationContent } from '../../../core/models/documentation';
+import { HttpClient } from '@angular/common/http';
+
+const documentationURL = './../../../../assets/docs/README.md';
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts:6]: `const documentationURL = './../../../../assets/docs/README.md';`
=>
const documentationURL = 'assets/docs/README.md';

Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.service.ts]: My bad

**Comment by hrishikeshpaul** [comment]
I think it would be nice to have the "about" panel expanded. What do you think?

**Comment by bherr2** [comment]
> I think it would be nice to have the "about" panel expanded. What do you think?

Makes sense. Perhaps with logic "always open the 'first' expansion panel"

**Comment by bherr2** [comment]
PS there is a merge conflict in projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html . I think you need to pull from develop. It looks like you've been pulling from `main` for PRs, but it needs to always come from/to `develop`.

**Comment by hrishikeshpaul** [comment]
> PS there is a merge conflict in projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html . I think you need to pull from develop. It looks like you've been pulling from main for PRs, but it needs to always come from/to develop.

I'll resolve that

**Comment by bherr2** [comment]
Tests are failing: https://github.com/hubmapconsortium/ccf-ui/runs/1937612168?check_suite_focus=true#step:10:518

**Comment by hrishikeshpaul** [comment]
Yeah, I'l trying to figure out how to do it

**Comment by bherr2** [comment]
You should also find a way to wait till the text is loaded before the dialog opens so that when it opens it doesn't jump (ie it immediately pops up with no content and then the whole div expands quickly once the data is loaded).

**Comment by hrishikeshpaul** [comment]
> You should also find a way to wait till the text is loaded before the dialog opens so that when it opens it doesn't jump (ie it immediately pops up with no content and then the whole div expands quickly once the data is loaded). 

Then I think I need to text processing function outside the dialog into the parent component

**Comment by bherr2** [comment]
> > You should also find a way to wait till the text is loaded before the dialog opens so that when it opens it doesn't jump (ie it immediately pops up with no content and then the whole div expands quickly once the data is loaded).
> 
> Then I think I need to text processing function outside the dialog into the parent component

Makes sense

**Comment by hrishikeshpaul** [comment]
Although, by doing so, there would be a small delay between click the button and the modal appearing.

**Comment by bherr2** [comment]
> Although, by doing so, there would be a small delay between click the button and the modal appearing.

The only other thing would be to have a fixed height on the modal so that it doesn't jump

**Comment by hrishikeshpaul** [comment]
Yeah, specifying a height at the beginning would be better

**Comment by bherr2** [comment]
Maybe both? I don't like that it is using the service directly. It is no longer a pure component. It'd be better if the info dialog took an array of "InfoSections" or whatever you're calling them.

**Comment by hrishikeshpaul** [comment]
Yes sounds good!

**Comment by hrishikeshpaul** [comment]
Okay I fixed the jumping behavior. I'm still trying to figure out the test cases. I know 3 of them are failing, so I'm trying to see how to fix it

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.ts [hunk 5]
```diff
@@ -11,15 +12,19 @@ import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
   styleUrls: ['./info-dialog.component.scss']
 })
 export class InfoDialogComponent {
+
+  documentationComtents: DocumentationContent[];
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.ts:16]: typo

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.scss [hunk 6]
```diff
@@ -34,5 +34,13 @@
       margin-top: 0;
       margin-bottom: 0.5rem;
     }
+
+    .panel {
+      margin-bottom: 1.5rem;
+    }
+
+    .dialog-content {
+      height: 50rem !important; min-height: 20px !important;
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.scss:43]: one line per style please. Also, is `!important` required here? If there is another way that's preferred.

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html [hunk 7]
```diff
@@ -5,73 +5,11 @@
...
-          <a class="link-blue" rel="noopener noreferrer" href="https://www.nlm.nih.gov/research/visible/visible_human.html" target="_blank">Visible Human Project</a>.
-        </p>
-
-        <p class="description">
-          The work is funded by NIH Award <a class="link-blue" rel="noopener noreferrer" href="https://projectreporter.nih.gov/project_info_description.cfm?aid=9687220" target="_blank">OT2OD026671</a>.
-        </p>
+      
+      <mat-dialog-content class="mat-typography dialog-content">
+        <div *ngFor="let content of documentationComtents; let i = index;" class="panel">
+          <ccf-info-panel [content]="content" [index]="i"></ccf-info-panel>
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html:11]: You sure you want to break this out as a separate component vs just using the little bit of code used in ccf-info-panel directly here?

**Review by bherr2 — CHANGES_REQUESTED** [review]
Much better without the jumping. I have a few changes (+ fixing the tests), but otherwise looks good.

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html (hunk 7 above)
Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.component.html]: Yeah you're right. I've refactored it

File: projects/ccf-rui/src/app/modules/info/info-button/info-button.component.ts [hunk 8]
```diff
@@ -17,17 +19,29 @@ export class InfoButtonComponent {
    * Creates an instance of info button component.
    *
    * @param dialog Reference to the dialog creation service.
+   * @param infoButtonService Reference to the info button service
    */
-  constructor(private readonly dialog: MatDialog) { }
+  constructor(private readonly dialog: MatDialog, private infoButtonService: InfoButtonService) {
+    infoButtonService.markdownContent.subscribe(data => {
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-button/info-button.component.ts:25]: add `ngOnDestroy` and unsubscribe from the observable

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.module.ts [hunk 9]
```diff
@@ -1,12 +1,16 @@
...
 import { MatDialogModule } from '@angular/material/dialog';
 import { MatIconModule } from '@angular/material/icon';
+import { MatExpansionModule } from '@angular/material/expansion';
+import { MarkdownModule } from 'ngx-markdown';
 
 @NgModule({
   declarations: [InfoDialogComponent],
-  imports: [CommonModule, MatDialogModule, MatIconModule],
+  imports: [CommonModule, MatDialogModule, MatIconModule, MatExpansionModule, MarkdownModule.forRoot({
+    sanitize: SecurityContext.NONE
```
Comment (bherr2) [inline projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.module.ts:12]: Curious, why is this necessary?

**Review by bherr2 — CHANGES_REQUESTED** [review]
One minor change, one question, and tests still need fixed.

**Comment by bherr2** [comment]
The http.get test errors are pretty easy to fix, just need to mock the http service (see @axdanbol ).
https://github.com/hubmapconsortium/ccf-ui/pull/489/checks?check_run_id=1955505397#step:10:541

**Comment by hrishikeshpaul** [comment]
Yeah I'm working on the tests, and trying to figure out mocking the HTTP service

File: projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.module.ts (hunk 9 above)
Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-dialog/info-dialog.module.ts:12]: So Angular automatically sanitizes the DOM when you're inputting elements in it (to prevent XSS) so without this, I was getting errors that didn't allow the parsing to happen. But turns out removing that now doesn't cause any errors. I'll remove this.

File: projects/ccf-rui/src/app/modules/info/info-button/info-button.component.ts (hunk 8 above)
Comment (hrishikeshpaul) [reply projects/ccf-rui/src/app/modules/info/info-button/info-button.component.ts:25]: Should the unsubscribe be when the modal is closed? The button still remains on the UI throughout

Comment (bherr2) [reply projects/ccf-rui/src/app/modules/info/info-button/info-button.component.ts:25]: No, just on destroy of the info-button (not likely to happen, but its a best practice to unsubscribe from all subscriptions when a component is destroyed).
